core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -754,20 +756,17 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
             }
         }
         if (source.innerHits() != null) {
-            XContentParser innerHitsParser = null;
-            try {
-                innerHitsParser = XContentFactory.xContent(source.innerHits()).createParser(source.innerHits());
-                innerHitsParser.nextToken();
-                this.elementParsers.get("inner_hits").parse(innerHitsParser, context);
-            } catch (Exception e) {
-                String sSource = "_na_";
+            for (Map.Entry<String, InnerHitBuilder> entry : source.innerHits().getInnerHitsBuilders().entrySet()) {
                 try {
-                    sSource = source.toString();
-                } catch (Throwable e1) {
-                    // ignore
+                    queryShardContext.reset();
```

Comment:
This is the same logic in `QueryShardContext#toQuery()` where we reset also twice. Personally I think a reset at the end is sufficient, but I kept the logic consistent with this method.

The reason we need to invoke reset at all here is because inner hits modifies the `QueryShardContext#nestedScope`, so we need to reset at the end.